sklearn_numba_dpex/kmeans/kernels/utils.py

```diff
@@ -25,6 +25,150 @@
 import numba_dpex as dpex
 
 
+@lru_cache
+def make_relocate_empty_clusters_kernel(
+    n_relocated_clusters,
+    n_features,
+    n_selected_gt_threshold,
+    work_group_size,
+    dtype,
+):
+    n_work_groups_for_cluster = math.ceil(n_features / work_group_size)
+    n_work_items_for_cluster = n_work_groups_for_cluster * work_group_size
+    global_size = n_work_items_for_cluster * n_relocated_clusters
+
+    n_selected_gt_threshold = n_selected_gt_threshold - 1
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    # fmt: off
+    def relocate_empty_clusters(
+        X_t,                        # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,              # IN READ-ONLY   (n_samples,)
+        assignments_idx,            # IN             (n_samples,)
+        samples_far_from_center,    # IN             (n_samples,)
+        empty_clusters_list,        # IN             (n_clusters,)
+        per_sample_inertia,         # INOUT          (n_samples,)
+        centroids_t,                # INOUT          (n_features, n_clusters)
+        cluster_sizes               # INOUT          (n_clusters,)
+            ):
+    # fmt: on
+        """NB: because of how the array was created, samples_far_from_center values
+        are located between (n_selected_gt_threshold - relocated_idx) and
+        (relocated_idx) indices.
+        """
+        group_idx = dpex.get_group_id(0)
+        item_idx = dpex.get_local_id(0)
+        relocated_idx = group_idx // n_work_groups_for_cluster
+        feature_idx = ((group_idx % n_work_groups_for_cluster) * work_group_size) + item_idx
+
+        if feature_idx >= n_features:
+            return
+
+        relocated_cluster_idx = empty_clusters_list[relocated_idx]
+        new_location_X_idx = samples_far_from_center[n_selected_gt_threshold - relocated_idx]
+        new_location_previous_assignment = assignments_idx[new_location_X_idx]
+
+        new_centroid_value = X_t[feature_idx, new_location_X_idx]
+        new_location_weight = sample_weight[new_location_X_idx]
+        X_centroid_addend = new_centroid_value * new_location_weight
+
+        # Cancel the contribution to the updated centroids of the sample that was once
+        # assigned to new_location_previous_assignment but is now assigned to the
+        # cluster of the centroids that relocates to this sample
```

Comment:
I find this comment a bit lengthy. Can this rephrased? For instance:

```suggestion
        # Cancel the contribution to new_location_previous_assignment, the 
        # previously updated centroids of the sample.
        # This contribution will be assigned to the centroid of the clusters
        # that relocates to this sample.
```

Note that I turned "the cluster of the centroids" to "the centroid of the clusters".